ccmlib/common.py

```diff
@@ -76,6 +77,36 @@ class UnavailableSocketError(CCMError):
     pass
 
 
+class LogPatternToVersion(object):
+
+    def __init__(self, versions_to_patterns, default_pattern):
+        self.versions_to_patterns = versions_to_patterns
+        if "0" not in self.versions_to_patterns:
```

Comment:
I think I need some clarification. 

The default pattern is a base case for our search through `versions_to_patterns`, i.e. when we can't find any version key less than the specified version, we need some answer. We need this because `max(...)` will blow up if we pass it an empty list as its iterable. In this case, if the default pattern isn't a part of `versions_to_patterns`, we will generate an empty list and blow up `max(...)`. If, in this same case, default is a property of the object, we have to catch the error `max(...)` is throwing and check if default exists.

I think that logic looks [something like this.](https://gist.github.com/smccarthy788/ece0ab0f4b9aa42ebb916c6a1ce5719a)

However, this implementation would preclude the default pattern from being `None`, which I think is the most correct answer in situations where a pattern didn't exist before some version. I suspect this is just an implementation error though. (actually, by the time I've posted this I think I've found a way to fix it, but it blows up `__repr__`)

I agree with you on `versions`, `__iter__`, and `__len__`, however, I use `patterns` in the `TimeoutError` implemented in `cluster.py`. I think `x.patterns` is more readable than `x.versions_to_patterns.values()`